### PR TITLE
39 book room

### LINKED ISSUE #39: Book a Room
### User story
As a user I am able to reserve a room
### Narrative
As a user I would like to reserve a room at a specific time. For the scope of this issue, assume no approvals are needed to book a room. 
### Out of scope
Limited to simply reserving a room. 
### Acceptance criteria

| # | Given | When | Then |
|---|-------|------|------|
|1|I am user of the system|I wish to book a room|I can book the room if its available|
|2|I am user of the system|I wish to book a room that is already booked|I cannot book the room|

### UI Mockup:
image

### LINKED ISSUE #234: Bug in update a room
## Expected Behavior
User tries to update a room
User changes a field in the update modal
User submits the form and the room is updated and displayed on the rooms list page

## Current Behavior
User tries to update a room
User changes a field in the update modal
User submits the form and receives a 404 error

## Possible Solution
Add the update route 

## Steps to Reproduce
1. Log in with valid credentials
2. Navigate to the rooms page
3. Select a room from the list below and click on update
4. Change one of the values and submit the form

## Context (Environment)
N/A

# Bug Submission Checklist
- [x] Replication steps
- [ ] Screenshots
- [ ] Tagged commits fixing the bug for traceability
- [ ] Tagged commits updating tests to cover the reported bug	
- [ ] (optional) tagged with “performance” label if performance related

### PR DESCRIPTION
**Pull request by jonathanp16**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**
Please provide enough information so that others can review your pull request:

Added the booking request model and create as well as update functionality. Also added a status field to the room model to differentiate between available and unavailable rooms. Fixed some tests that were failing for rooms due to the newly added field. Users can now create booking requests for a specific room. Also fixed bug where user update form did not display the current values (#234 )

### Test plan (required)
npm run test
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s|
|---|-------|------|------|------|------|
| BookingRequestsList.vue            |   77.78 |      100 |      40 |   77.78 | 130-141|
 | CreateBookingRequestForm.vue       |     100 |      100 |     100 |     100 |
 | Index.vue                          |     100 |      100 |     100 |     100 |
 | UpdateBookingRequestForm.vue       |      95 |    81.25 |     100 |      95 | 125|

### Closing issues
List all issues this PR resolves.
- closes #234  
- closes #39 

### Pre-Merge checklist 
Complete these items before requesting reviews
- [ ] Static Analysis Rules Validated
- [ ] All commits follow naming conventions 

### Sub-tasks Checklist
- [x] UI Implementation
- [x] UI Test
- [x] Backend Implementation
- [x] Backend Tests
- [ ] Integration Tests
- [ ] System Tests

### CODE REVIEW

File: resources/js/Components/Navbar.vue [hunk 1]
```diff
@@ -0,0 +1,112 @@
+<template>
```
Comment (Alex-Hayeur) [inline resources/js/Components/Navbar.vue:1]: Can you go into more detail for what this functionality is for? It doesn't seem to relate specifically to the Booking feature. Will this cause conflicts with the navbar/theme tasks?

**Review by Alex-Hayeur — COMMENTED** [review]
There are some bugs that may need to be addressed and the one comment I'd like to be answered to make sure this PR remains scoped in one one feature, but the general functionality looks solid.

File: resources/js/Components/Navbar.vue (hunk 1 above)
Comment (jonathanp16) [reply resources/js/Components/Navbar.vue:1]: This was just used to test the dynamic tables that I created. Forgot to clean that up but will remove it now.

**Review by Alex-Hayeur — CHANGES_REQUESTED** [review]
Alright, removing the test navbar is fine. 
The bugs are listed as major, but are all just missing descriptions on self-descriptive tables and tags. It's best-practice in general to add them but not sure if that's important to our own project's practices.

The Code Coverage % is a little low, but on further examination, it's caused by empty classes which are going to be filled later (show and edit booking requests)
However it looks like there's a test missing for deleting a booking request. If that feature is in scope for this issue, could you add it?

**Review by Alex-Hayeur — APPROVED** [review]
We discussed, the delete function isn't implemented otherwise + is out of scope of the issue. It looks good to merge.

File: app/Models/Room.php [hunk 2]
```diff
@@ -18,6 +18,15 @@ class Room extends Model
...
         'floor',
-        'building'
+        'building',
+        'status'
     ];
+
+     /**
+     * Get the booking requests for the room.
+     */
+    public function bookingrequests()
```
Comment (alecadub) [inline app/Models/Room.php:28]: CamelCase for function name

File: app/Models/User.php [hunk 3]
```diff
@@ -60,4 +60,12 @@ class User extends Authenticatable
     protected $appends = [
         'profile_photo_url',
     ];
+
+    /**
+     * Get all booking requests created by this user.
+     */
+    public function bookingrequests()
```
Comment (alecadub) [inline app/Models/User.php:67]: CamelCase for function name

File: database/seeders/RoomSeeder.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+     * @return void
+     */
+    public function run()
+    {
+        Room::create([
+            'name' => Str::random(10),
+            'number' => Str::random(10),
+            'floor' => rand(1, 25),
+            'building' => Str::random(10),
+            'status' => Str::random(5)
```
Comment (alecadub) [inline database/seeders/RoomSeeder.php:23]: This gives a random status, but status are only available or unavailable, should use same logic as the factory.

File: resources/js/Pages/Admin/Rooms/RoomsList.vue [hunk 5]
```diff
@@ -23,8 +23,8 @@
                             <div class="text-md mx-3">Building</div>
```
Comment (alecadub) [inline resources/js/Pages/Admin/Rooms/RoomsList.vue:23]: Should add status here !

**Review by alecadub — CHANGES_REQUESTED** [review]
Small changes to be made in the code, and you should refactor to fix the sonar cloud bugs! the rest looks good! (also rebase please!)

However, have some concerns:

1. We see the list of available and unavailable rooms for booking request, is that expected ? Should it be on available ones
image

2. I'm not a fan at all of updating the booking request with user or booking id :/ it should be a dropdown with the available choices no ? As soon as I enter an ID that doesn't exists, the application crash!
image

3. As mentioned by @Alex-Hayeur , the delete function doesn't work but it's out of scope.

**Review by alecadub — APPROVED** [review]

**Review by Alex-Hayeur — APPROVED** [review]
Passing approval now that requested changes have been made.

**Comment by alexstojda** [comment]
Taking a look at this conflict 👀

File: database/migrations/2020_11_21_160732_create_rooms_table.php [hunk 6]
```diff
@@ -19,6 +19,7 @@ public function up()
             $table->string('number');
             $table->integer('floor');
             $table->string('building');
+            $table->string('status');
```
Comment (WindOfRussia) [inline database/migrations/2020_11_21_160732_create_rooms_table.php:22]: can we move this to a separate migration file?

**Review by Alex-Hayeur — CHANGES_REQUESTED** [review]
After team discussion, I confirmed that the changes made to the original migration need to be undone.
Migrations need to be made as separate files. You can refer to this guide 
https://www.tutsmake.com/laravel-migration-add-a-column-to-an-existing-table/

**Review by WindOfRussia — CHANGES_REQUESTED** [review]

**Review by Alex-Hayeur — DISMISSED** [review]
Looks hood with the new migration file

**Review by alecadub — APPROVED** [review]

**Review by WindOfRussia — DISMISSED** [review]
good with the new migration

**Comment by Alex-Hayeur** [comment]
The unused code in the booking request controllers (For out-of-scope functions like delete, show and edit) is causing the test coverage to drop below 80%

https://sonarcloud.io/component_measures?id=CSU-Booking-Platform_application&metric=new_coverage&pullRequest=235&selected=CSU-Booking-Platform_application%3Aapp%2FHttp%2FControllers%2FBookingRequestController.php&view=list

To get it to pass you'll need to either remove the unused code or write tests for it.

**Review by alecadub — APPROVED** [review]

**Review by Alex-Hayeur — APPROVED** [review]
Code coverage is appropriate, and the rest still looks good.

It's good to merge.

**Review by alexstojda — APPROVED** [review]

**Review by Alex-Hayeur — APPROVED** [review]
passing approval over from before resolving merge conflicts

**Review by alexstojda — APPROVED** [review]

**Review by Alex-Hayeur — APPROVED** [review]

**Comment by alexstojda** [comment]
Seems github is having a bad day with this PR, resorting to manual merge

**Comment by alexstojda** [comment]
This PR gave me nightmares, so I made a new one: https://github.com/CSU-Booking-Platform/application/pull/265

**Review by danseremet — APPROVED** [review]
Looks good to me

**Pull request closed without merging**